lib/ui/window.dart

```diff
@@ -365,11 +407,95 @@ class Window {
   /// Message handlers must call the function given in the `callback` parameter.
   /// If the handler does not need to respond, the handler should pass `null` to
   /// the callback.
-  PlatformMessageCallback onPlatformMessage;
+  PlatformMessageCallback get onPlatformMessage => _onPlatformMessage;
+
+  set onPlatformMessage(PlatformMessageCallback callback) {
+    _onPlatformMessage = _zonedPlatformMessageCallback(callback);
+  }
+
+  PlatformMessageCallback _onPlatformMessage;
 
   /// Called by [_dispatchPlatformMessage].
   void _respondToPlatformMessage(int responseId, ByteData data)
       native "Window_respondToPlatformMessage";
+
+  /// Wraps the given [callback] in another callback that ensures that the
+  /// original callback is called in the zone it was registered in.
+  static VoidCallback _zonedVoidCallback(VoidCallback callback) {
```

Comment:
The way you've implemented this we're already paying the cost to track the registration zone, so the solution is probably just to have the setters remember the zone in a field next to the callback, and then in hooks.dart et al when we call the callbacks just have them call the callbacks with runUnaryGuarded et al themselves. It'd probably be cheaper, too, since you wouldn't have to track the new closure.